### PR TITLE
Add ability to create Http Imposter with Proxy Stub

### PR DESCRIPTION
**Pull request by tjgould1983**

### CODE REVIEW

**Comment by mattherman** [comment]
Thanks for the pull request! I'll try to take a look sometime today or tomorrow.

**Comment by tjgould1983** [comment]
I am aware i need to add a test or two :)

**Comment by mattherman** [comment]
Still haven't had much time to look at this. I'm still going to try to get it done before the end of the weekend though.

**Comment by tjgould1983** [comment]
Any luck looking at this pr? Please

File: MbDotNet/Models/Imposters/Imposter.cs [hunk 1]
```diff
@@ -1,29 +1,35 @@
...
+        /// Optional name for the imposter, used in the logs.
+        /// </summary>
+        [JsonProperty("name")]
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// Adds mock verification support by remembering the requests made to this imposter. Note that this represents a memory leak for any long running mb process, as requests are never forgotten.
+        /// </summary>
+        [JsonProperty("recordRequests")]
+        public bool RecordRequests { get; set; }
```
Comment (mattherman) [inline MbDotNet/Models/Imposters/Imposter.cs]: Why was this added? I don't see this on the [imposters contract](http://www.mbtest.org/docs/api/contracts) and it doesn't appear to be used anywhere. Unless there is a different reason for it, please remove it.

File: MbDotNet/Models/Responses/Fields/ProxyResponseFields.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        [JsonProperty("to", NullValueHandling = NullValueHandling.Ignore)]
+        public Uri To { get; set; }
+
+        [JsonProperty("mode", NullValueHandling = NullValueHandling.Ignore)]
+        [JsonConverter(typeof(StringEnumConverter))]
+        public ProxyMode Mode { get; set; }
+
+        [JsonProperty("predicateGenerators", NullValueHandling = NullValueHandling.Ignore)]
+        public List<PredicateBase> PredicateGenerators { get; set; }
```
Comment (mattherman) [inline MbDotNet/Models/Responses/Fields/ProxyResponseFields.cs:21]: I'm not sure using `PredicateBase` here will work. Looking at the [documentation around predicate generators](http://www.mbtest.org/docs/api/proxies#proxy-predicate-generators), it looks like the generators always have a `matches` field. Since that seems to be identical to a regular matches predicate, we should be able to make this a `List<MatchesPredicate>` instead.

Also, please make it an `IList` instead of `List`.

File: MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+﻿using System.Collections.Generic;
+using MbDotNet.Enums;
+using Newtonsoft.Json;
+
+namespace MbDotNet.Models.Predicates.Fields
+{
+    public class PredicateGeneratorFields : PredicateFields
```
Comment (mattherman) [inline MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs:7]: I don't think this type is going to be enough. It is specific to the predicate fields of an HTTP imposter predicate, but we'll need to support TCP as well (see `TcpPredicateFields`).

I'm not very familiar with proxies in mountebank, but the documentation appears to show that you can use the regular predicate fields with values like this:
```
"predicateGenerators": [{
    "matches": {
        "query": { "q": "mountebank" }
    }
}]
```

Or the boolean version that you are supporting here:
```
"predicateGenerators": [{
    "matches": { "query": true },
    "caseSensitive": true
}]
```

I think we need to find a way to have HTTP and TCP value predicate fields and HTTP and TCP boolean predicate fields. It's something I've needed in the past in order to implement things like the [exists predicate](http://www.mbtest.org/docs/api/predicates#predicates-exists), but never came up with a good way of doing it. One concern is that I don't think we would want people to be creating the other predicate types, like `EqualsPredicate` with the boolean version of the fields since it doesn't really make sense.

I'll give this some more thought, let me know if you have any ideas on how we can represent both the value-style and boolean-style predicate fields while still restricting which types of predicates can use which predicate field types.

File: MbDotNet/Models/Stubs/HttpStub.cs [hunk 4]
```diff
@@ -123,13 +124,36 @@ public HttpStub Returns(HttpStatusCode statusCode, IDictionary<string, string> h
...
+        }
+
+        /// <summary>
+        /// Adds a proxy to the stub with the specified content type
+        /// </summary>
+        /// <param name="to">endpoint address to proxy to</param>
+        /// <param name="proxyMode">proxyalways, proxyonce or proxytransparent</param>
+        /// <param name="predicateGenerators">list of predicates that a proxy repsonse will be recorded for</param>
+        /// <returns></returns>
+        public HttpStub Returns(Uri to, ProxyMode proxyMode, List<PredicateBase> predicateGenerators)
```
Comment (mattherman) [inline MbDotNet/Models/Stubs/HttpStub.cs:137]: Maybe change this method to `ReturnsProxy` or something similar to differentiate it from the `Returns` methods that will give back specific content.

File: MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs (hunk 3 above)
Comment (mattherman) [reply MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs:7]: I guess worst-case scenario we just define the two styles of predicate fields and don't restrict them, and if someone uses boolean fields on an equals predicate or something it just won't work. I'd rather restrict it somehow though so people can't create invalid predicates.

File: MbDotNet/Models/Responses/ProxyResponse.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+using MbDotNet.Models.Responses.Fields;
+
+namespace MbDotNet.Models.Responses
+{
+    public class ProxyResponse<T> : ResponseBase where T : ResponseFields, new()
```
Comment (mattherman) [inline MbDotNet/Models/Responses/ProxyResponse.cs:6]: Please add a test for this. It's not super helpful, but it will at least maintain consistency with the other response classes. See `IsResponseTests`.

File: MbDotNet/Models/Imposters/Imposter.cs (hunk 1 above)
Comment (tjgould1983) [reply MbDotNet/Models/Imposters/Imposter.cs]: Was trying to make use of the new property - see http://www.mbtest.org/docs/protocols/http. Will double check

Comment (mattherman) [reply MbDotNet/Models/Imposters/Imposter.cs]: Ah, I see. It must not have been added to that contracts page. I think I'd prefer if that was added in a separate pull request since it is not really related to the proxy changes.

File: MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs (hunk 3 above)
Comment (mattherman) [reply MbDotNet/Models/Predicates/Fields/PredicateGeneratorFields.cs:7]: Also, I guess you don't need to add the implementations for TCP as long as the final design is able to support adding it. I can always add it later.

**Comment by mattherman** [comment]
Are you still working on this? Do you have any questions about the feedback I provided?

**Comment by mattherman** [comment]
I have some spare time today so I think I will merge this as is and then make the changes I suggested. Thanks again for the contribution!

**Comment by tjgould1983** [comment]
Sorry been busy at work with other things. Thanks for implementing properly

**Comment by mattherman** [comment]
No problem. If you're curious where it ended up, I just merged the rest of the changes in #48 .
